### PR TITLE
Add multi-backend testing for Keystone SAML Mellon

### PR DESCRIPTION
**Pull request by ionutbalutoiu**
The new updated tests rely on new testing bundles deployed with
two local IdPs via the Juju charm https://jaas.ai/u/ionutbalutoiu/test-saml-idp.

### CODE REVIEW

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:30]: Changing the name from `CharmKeystoneSAMLMellonTest` to `CharmKeystoneSAMLMellonIDP1Test` will, I think, break stable versions of the charm tests.  We tend to have to leave the old names alone and add new ones only so as to not break existing stable charms.

**Review by ajkavanagh — CHANGES_REQUESTED** [review]
Looks good apart from the test name change for stable charms / existing tests (e.g. from a tests.yaml).  As the Test names are essentially the API to the testing library they tend to be "forever".

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:30]: @ajkavanagh

Makes sense. I'll update the PR, and make sure to keep the backwards compatibility.

**Comment by ionutbalutoiu** [comment]
> Looks good apart from the test name change for stable charms / existing tests (e.g. from a tests.yaml). As the Test names are essentially the API to the testing library they tend to be "forever".

I updated the PR and took care of the backwards compatibility. I also validated these changes with a `func-smoke` against:
* Latest master with my bundles changes from https://review.opendev.org/c/openstack/charm-keystone-saml-mellon/+/770056
* The `stable/20.10` charm release.

Everything went fine on my local Zaza environment.

There's some duplicate code in the `zaza/openstack/charm_tests/saml_mellon/tests.py` between the old `CharmKeystoneSAMLMellonTest` class (which uses upstream `samltest.id`), and the new `BaseCharmKeystoneSAMLMellonTest` class (which uses the local `test-saml-idp`).

I left it like this on purpose, since we can remove the old testing class, once we fully drop the `samltest.id` dependency (even on stable charm releases).

**Review by AurelienLourot — CHANGES_REQUESTED** [review]
I haven't reviewed yet, just posting this comment first, but I'll circle back to it soon, thanks!

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:159]: @AurelienLourot Sounds good. I'll wait for your review, and address all the comments at once.

Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:161]: Added a comment about the deprecation of the old testing class using `samltest.id` dependency.

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (AurelienLourot) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:174]: This note was useful, can we keep it?

```
        # Note: The BaseKeystoneTest class sets the application_name to
        # "keystone" which breaks keystone-saml-mellon actions. Explicitly set
        # application name here.
```

**Review by AurelienLourot — CHANGES_REQUESTED** [review]
LGTM in general, thanks!

File: zaza/openstack/charm_tests/saml_mellon/setup.py
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/setup.py:183]: There's a pattern developing here (this is repeated 3 times).  Would breaking out a helper that does this and attaches the resource (or even add a new helper), help with the clarity and intention of the code.  e.g.

```python
zaza.utils.attach_resource(app, "idp-metadata', idp_metadata, ".xml")
```

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:188]: I tend to prefer (but don't change it if you prefer the current format):

```python
ip = self.vip if self.vip else unit.public_address
```

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:193]: This is test code.  Why not:

```python
self.assertNotIn("failed", action.data["status"], msg="The action failed ....")
```

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:199]:
```python
self.assertIn(ip, item[1])
```

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:205]: As above.

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:229]: One line if-else possible if preferred.

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:296]: This is a big function to be embedded in another.  Please break it out and add `@staticmethod` to it, plus docstring with purpose, parameters and types.

**Review by ajkavanagh — CHANGES_REQUESTED** [review]
Please see the single comments; apologies for not starting a review.   I'd already proceeding with the single comments and it was too late by then!

File: zaza/openstack/charm_tests/saml_mellon/setup.py
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/setup.py:140]: Yes. I'll add the commend.

File: zaza/openstack/charm_tests/saml_mellon/setup.py
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/setup.py:174]: In fact, I only need to run the `get-idp-metadata` action, and I noticed that we have `run_action_on_leader` and I don't even need to fetch an unit. I'll use that.

File: zaza/openstack/charm_tests/saml_mellon/setup.py
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/setup.py:183]: Good suggestion. Will implement it.

File: zaza/openstack/charm_tests/saml_mellon/setup.py
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/setup.py:203]: Same solution as above. I can use `runa_action_on_leader`.

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:174]: Will add it back.

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:188]: I like this as well. I'll change it.

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:193]: Will change it.

File: zaza/openstack/charm_tests/saml_mellon/tests.py
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:296]: This function was carried on from the previous implementation.

I agree with you. I'll clean it up.

**Comment by ionutbalutoiu** [comment]
@ajkavanagh @AurelienLourot

Thank-you for the reviews. I addressed all the commends. Please check the updated PR when you have a chance.

**Review by ajkavanagh — APPROVED** [review]
LGTM - thanks for making the  changes, particularly the code-gardening around the embedded function; much clearer now.

**Review by AurelienLourot — APPROVED** [review]
Perfect, thanks! Happy to land this when the corresponding gerrit review turns green. ( https://review.opendev.org/c/openstack/charm-keystone-saml-mellon/+/770056 )

**Comment by AurelienLourot** [comment]
Note: the `.decode()` call has been moved, this is what has changed in the last push-force. Still LGTM. @ionutbalutoiu can you please not push-force? Because we can't easily spot here in the browser what changed and this requires special steps to figure it out. Thanks!

**Comment by ionutbalutoiu** [comment]
> Note: the `.decode()` call has been moved, this is what has changed in the last push-force. Still LGTM. @ionutbalutoiu can you please not push-force? Because we can't easily spot here in the browser what changed and this requires special steps to figure it out. Thanks!

Oh, sorry about `force-push`. I tend to have a single clean commit. But I realize that it can get confusing when the changes are reviewed already.

The `decode()` had to be removed from there, because the helper function is meant to get a `str` (not an encoded `str`).

After I got your approvals, I quickly ran a local Zaza run and noticed that the function was failing when it got `str`, because it tried to `decode()` it. The `decode()` needs to be called before using the helper (if needed).

That's the full story of the most recent change. Our local Zaza run went through without any problems after fixing this.

**Review by AurelienLourot — APPROVED** [review]

**Pull request merged**
